### PR TITLE
Allocate token's content into an arena

### PR DESCRIPTION
**Pull request by maciektr**
commit-id:5e4b122f

Add token stream ser/de tests

---

**Stack**:
- #1749
- #1748
- #1745
- #1747
- #1722 ⬅

⚠️ *Part of a stack created by [spr](https://github.com/ejoffe/spr). Do not merge manually using the UI - doing so may have unexpected results.*

### CODE REVIEW

File: plugins/cairo-lang-macro/src/types/token.rs [hunk 1]
```diff
@@ -34,13 +109,118 @@ pub struct TextSpan {
...
+
+impl Hash for InternedStr {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.as_ref().hash(state);
+    }
+}
+
+/// This wrapper de-allocates the underlying buffer on drop.
+#[derive(Debug, Default)]
+pub struct BumpWrap(pub Bump);
```
Comment (mkaput) [inline plugins/cairo-lang-macro/src/types/token.rs:191]: why is this `pub`?

File: plugins/cairo-lang-macro/src/types/token.rs
Comment (mkaput) [inline plugins/cairo-lang-macro/src/types/token.rs:1]: doc everything

**Review by mkaput — CHANGES_REQUESTED** [review]

**Review by mkaput — APPROVED** [review]

File: plugins/cairo-lang-macro/src/lib.rs [hunk 2]
```diff
@@ -18,17 +18,22 @@
...
 };
 use std::ffi::{c_char, CStr, CString};
+use std::ops::Deref;
 
 mod types;
 
 pub use types::*;
 
+// A thread-local allocation context for allocating tokens on proc macro side.
+thread_local!(static CONTEXT: RefCell<AllocationContext> =  RefCell::default() );
```
Comment (wawel37) [inline plugins/cairo-lang-macro/src/lib.rs:35]: Could you explain a little bit why do we have to make it `thread_local`?

Comment (Draggu) [reply plugins/cairo-lang-macro/src/lib.rs:35]: `RefCell` is not `Sync`, this makes access to it easier, if it is not `thread_local` simple `RefCell<T>` must be replaced with `Arc<Mutex<T>>`

Comment (maciektr) [reply plugins/cairo-lang-macro/src/lib.rs:35]: Exactly this ^

We want this to be available globally, so `Token::new` can access it implicitly.

File: plugins/cairo-lang-macro/src/types/token.rs [hunk 3]
```diff
@@ -1,17 +1,85 @@
...
     pub metadata: TokenStreamMetadata,
 }
 
+/// This module implements deserialization of the token stream, for the serde feature.
+/// This is intermediate representation is needed, as real [`Token`] only contains a reference to the
+/// represented string, which needs to be allocated outside the [`Token`] struct.
+/// Here we allocate each token to an owned String with SerDe and then copy it's content into context.
+#[cfg(feature = "serde")]
+#[doc(hidden)]
+mod de {
```
Comment (wawel37) [inline plugins/cairo-lang-macro/src/types/token.rs:25]: nit: maybe
```suggestion
mod deserializer {
```
?

**Review by wawel37 — APPROVED** [review]
As for my knowledge it looks really good :)

**Pull request merged**